Cython/Compiler/Nodes.py

```diff
@@ -4962,14 +4962,22 @@ def generate_execution_code(self, code):
             code.putln("else {")
         else:
             code.putln("else if (")
-            code.putln("#if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY")
-            code.putln(f"unlikely(Py_TYPE({self_arg})->tp_dictoffset != 0)")
+            code.putln("#if !(CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY)")
+            # If CYTHON_USE_TYPE_SPECS then all extension types are heap-types so the check below automatically
+            # passes and thus takes the slow route.
+            # Therefore we do a less thorough check - if they type hasn't changed then clearly it hasn't
+            # been overridden, and if the type isn't GC then it also won't have been overridden.
+            if self.py_func.entry.scope.lookup_here("__dict__"):
+                code.putln("(1)")
+            else:
+                code.putln(f"unlikely(Py_TYPE({self_arg}) != "
+                           f"{self.py_func.entry.scope.parent_type.typeptr_cname} &&")
+                code.putln(f"__Pyx_PyType_HasFeature(Py_TYPE({self_arg}), Py_TPFLAGS_HAVE_GC))")
             code.putln("#else")
-            dict_str_const = code.get_py_string_const(EncodedString("__dict__"))
-            code.putln(f'PyObject_HasAttr({self_arg}, {dict_str_const})')
+            code.putln(f"unlikely(Py_TYPE({self_arg})->tp_dictoffset != 0 || "
+                       f"__Pyx_PyType_HasFeature(Py_TYPE({self_arg}), (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))")
```

Comment:
This is the existing behaviour but it does have a problem:

```
cdef class C:
    cpdef f(self):
        return 0

    def call_f(self):
        return self.f();

cdef class D(C):
    def f(self):
        return 1
```

`D().call_f()` will return `0` while if `D` was a regular class then it would be overridden. The type-specs version behaves differently and uses the override from `D`. I think it's hard to avoid that difference in behaviour.

Reply:
So, if we're in a subclass, we'd have to always do the dict lookup. That seems quite a performance penalty.

Can't we warn about the case of overriding `cpdef` with `def`? That seems unintentional and risky.

Reply:
The warning already exists:

```
warning: testwarning.pyx:9:4: Overriding cdef method with def method.
```

Reply:
Then I think we can make it clearer why this is an issue. Proposal:
"Overriding a c(p)def method with a def method. This can lead to different methods being called depending on the call context. Consider using a cpdef method for both."

Reply:
Updated.

Reply:
There's a `w_cdef_override` errors test that probably needs adjusting.